### ISSUE TITLE
[HOLD for payment 2024-12-16] [HOLD for payment 2024-12-11] [$250] [Held requests] Partial paying a report causes the wrong amounts to show in the LHN

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#48760]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.30-15
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @JmillsExpensify 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1725649428907579

## Action Performed:
1. On a Collect policy, have user A submit several expenses to User B
2. As User B, hold at least two expenses, approve the rest
## Expected Result:
The report amount in the LHN is correct.
## Actual Result:
The report amount in the LHN only shows the amount of one expense
## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

![CleanShot 2024-09-06 at 20 44 35@2x](https://github.com/user-attachments/assets/9e03dff4-d748-4700-850d-33e675785926)

![CleanShot 2024-09-06 at 20 45 06@2x](https://github.com/user-attachments/assets/e938df51-7a84-40bd-9c1d-be1e9fef6d36)

https://github.com/user-attachments/assets/d0f83087-4e76-4ac9-b13f-1c07f397d79f

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021833144594407206391</li>
        <li>Upwork Job ID: 1833144594407206391</li>
        <li>Last Price Increase: 2024-09-09</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @robertjchen</details>

**Comment by jaydamani** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

wrong amount is show when some expenses of a report are held

### What is the root cause of that problem?

The [getMoneyRequestSpendBreakdown](https://github.com/Expensify/App/blob/7ebfaa0daaa43a20083250e0f263964ed5b647b2/src/libs/ReportUtils.ts#L2638) shows total of expenses which includes held expenses.

https://github.com/Expensify/App/blob/7ebfaa0daaa43a20083250e0f263964ed5b647b2/src/libs/ReportUtils.ts#L2649

### What changes do you think we should make in order to solve the problem?


Update [getMoneyRequestSpendBreakdown](https://github.com/Expensify/App/blob/7ebfaa0daaa43a20083250e0f263964ed5b647b2/src/libs/ReportUtils.ts#L2638) to use `unheldTotal` when available as it does not include held expenses.
So the new code will be

```js
let totalSpend = moneyRequestReport.unheldTotal ?? moneyRequestReport.total ?? 0;
```

**Comment by jaydamani** [comment]
Is there any place in the app where we show total like this and explicitly want to include held expenses?

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Approving partly a report with held transaction will show the incorrect amount for the new report that contains all the held transaction.

### What is the root cause of that problem?
There are 2 problems. First, when we approve the report, we create the new expense report optimistically that contains all the held transactions, but, we only use the first amount as the report total amount.
https://github.com/Expensify/App/blob/03ec468bfb14265d8953a8cfad7a0a875991fd8f/src/libs/actions/IOU.ts#L6390-L6394

Also, we don't optimistically update the approved report total by subtracting the current total with the new expense report total
https://github.com/Expensify/App/blob/03ec468bfb14265d8953a8cfad7a0a875991fd8f/src/libs/actions/IOU.ts#L6943-L6949

Second, even though the optimistic data is incorrect, we receive the correct data from the BE response. However, the LHN last message isn't updated immediately. The LHN shows the total from the expense report and we get the report inside the function instead of connecting the LHN component with the report.
https://github.com/Expensify/App/blob/03ec468bfb14265d8953a8cfad7a0a875991fd8f/src/libs/OptionsListUtils.ts#L625-L626

And even though we connect all reports inside lhn list,
https://github.com/Expensify/App/blob/25cdaa9be8520b6b633e2515c665a12234dcf970/src/components/LHNOptionsList/LHNOptionsList.tsx#L35

the OptionRowLHNData is memoized, so it won't re-render if the props are not changed.

### What changes do you think we should make in order to solve the problem?

Update the optimistic data to set the correct amount for the new expense report.
https://github.com/Expensify/App/blob/03ec468bfb14265d8953a8cfad7a0a875991fd8f/src/libs/actions/IOU.ts#L6394

We can calculate it from the report total and unheldTotal,
`(iouReport.total - iouReport.unheldTotal) * -1,`

or sum the amount from the individual transaction amount.
`holdTransactions.reduce((acc, transaction) => acc + transaction.amount, 0)`

Then, we need to update the current expense report total too.
https://github.com/Expensify/App/blob/03ec468bfb14265d8953a8cfad7a0a875991fd8f/src/libs/actions/IOU.ts#L6918-L6922
https://github.com/Expensify/App/blob/03ec468bfb14265d8953a8cfad7a0a875991fd8f/src/libs/actions/IOU.ts#L6943-L6949
```
...expenseReport,
total,
...,
```

Solving the optimistic data is currently enough for the LHN to show the correct amount. But if we also want to fix the 2nd issue, then we can connect the expense report in `OptionRowLHNData`.
```
const [iouReport] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT}${fullReport?.iouReportID}`);
```

Then, pass it to `SidebarUtils.getOptionData` > `OptionsListUtils.getLastMessageTextForReport` and replace the iouReport here with the new iouReport params.
https://github.com/Expensify/App/blob/03ec468bfb14265d8953a8cfad7a0a875991fd8f/src/libs/OptionsListUtils.ts#L625-L626

However, there is a BE issue, specifically the ApproveMoneyRequest API where the response returns the previous iouReportID instead of the new one. For example, let's say the current iouReportID is 1 which has 2 held transactions. After approving it, we optimistically move the 2 held transactions to a new report with an ID of 2 then update iouReportID to 2. But the ApproveMoneyRequest response revert the iouReportID back to 1. If we refresh/reopen the chat report, the OpenReport will return the correct iouReportID, 2. So, we need to fix the BE if we want to apply this solution.

If we don't want to rely on iouReportID, then we can follow the same logic here by taking the last action and extract the `iouReportID`.
https://github.com/Expensify/App/blob/03ec468bfb14265d8953a8cfad7a0a875991fd8f/src/libs/OptionsListUtils.ts#L626

Fortunately, we already get the `iouReportID` here,
https://github.com/Expensify/App/blob/25cdaa9be8520b6b633e2515c665a12234dcf970/src/components/LHNOptionsList/LHNOptionsList.tsx#L131

so we just need to retrieve the report and pass it down to OptionRowLHNData.
```
const itemIouReport = reports?.[`${ONYXKEYS.COLLECTION.REPORT}${iouReportIDOfLastAction}`];
...

iouReport={itemIouReport}
```

**Comment by trjExpensify** [comment]
Adding Robert for the secondary internal review of the proposals for hold. @s77rt can you review the above from Bernhard? Thanks!

**Comment by trjExpensify** [comment]
Just a request that we make sure we test proposed solutions with both iouReports and expenseReports, because they are different report types.

**Comment by s77rt** [comment]
@bernhardoj Thanks for the proposal. Can you explain how this is related to optimistic data? Per the OP video, the incorrect data is on User A and not User B. (The optimistic data you referred to are for User B)

Seems like a pusher related issue

**Comment by s77rt** [comment]
@bernhardoj Ah I see, we have multiple similar bugs. The RCA and the solution for the bug you mentioned make sense to me. But let's not update the `total`, as that total is meant for the total that was approved and not for what's left as total to be paid. Fixing `getReportFromHoldRequestsOnyxData` is sufficient.

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/48760#issuecomment-2337720896)

**Comment by s77rt** [comment]
@robertjchen Please note that we have two bugs reported here. The bug reported in the "Action Performed" section is FE and @bernhardoj's solution above will fix it. However the bug showcased in the OP video is a little different and seems related to pusher events. After the payer approved some expenses, the payee got wrong new amount that the WS still owes (it should be non-zero since some expenses are still held)

**Comment by kevinksullivan** [comment]
This one's moving forward. Looping in another BZ member as I'll be off till September 24

**Comment by trjExpensify** [comment]
@robertjchen are you happy with this proposal then? If so, can we assign @bernhardoj to move forward with it?

**Comment by robertjchen** [comment]
Yep, let's do it! 👍

**Comment by trjExpensify** [comment]
Great, @bernhardoj please proceed with a PR ASAP. Thanks!

**Comment by bernhardoj** [comment]
PR is ready

cc: @s77rt

**Comment by robertjchen** [comment]
PR reviewed and merged 👍

**Comment by robertjchen** [comment]
Potential BE fix in review ^

**Comment by robertjchen** [comment]
BE fix merged and being deployed today

**Comment by robertjchen** [comment]
I think we're all set here! The only remaining issue involves seeing a `0.00` amount in the LHN, but we'll be addressing it in a separate issue here: https://github.com/Expensify/App/issues/48093

**Comment by robertjchen** [comment]
cc: @RachCHopkins for next steps here!

**Comment by RachCHopkins** [comment]
Actually I think it's back to @kevinksullivan since it's after the 24th! I've just been babysitting!

But let me confirm the payments since the automation is borked. 

- Contributor: @bernhardoj to be paid $250 via newdot manual request
- Contributor+: @s77rt  paid $250 via newdot manual request

Upwork job [here](https://www.upwork.com/jobs/~021833144594407206391)

Is that correct @robertjchen?

**Comment by RachCHopkins** [comment]
And do we need @s77rt to complete the checklist? 

The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] The PR that introduced the bug has been identified. Link to the PR:
- [ ] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] Link the GH issue for creating/updating the regression test once above steps have been agreed upon

**Comment by bernhardoj** [comment]
Requested in ND.

**Comment by s77rt** [comment]
@bernhardoj I just found out we sum up the hold transactions values without taking into account their currency. This causes wrong value calculation if we hold multiple expenses with different currencies


https://github.com/user-attachments/assets/866ed0fd-6df2-43b0-bbc0-4cf0db182782

cc @robertjchen since you are working on fixing a similar bug on BE, I think you may face the same issue

**Comment by bernhardoj** [comment]
Oh, that's bad. That means we can't use the solution to iterate over the transaction amount. I think we need to revisit the approach of subtracting total - unheldTotal.

**Comment by s77rt** [comment]
@bernhardoj Will that approach be enough to make it work? I see this line that gets the currency based on the first hold transaction https://github.com/Expensify/App/blob/b5a0a2967635d117506e3dc4c1b78139d0d17b44/src/libs/actions/IOU.ts#L6457

If the first transaction currency is different than the others then the problem will still occur

**Comment by JmillsExpensify** [comment]
$250 approved for @bernhardoj

**Comment by bernhardoj** [comment]
Oh, I didn't notice that. I think we should use whatever the currency of the current IOU report, so it would be,
```
iouReport.currency,
```

What do you think?

**Comment by s77rt** [comment]
@bernhardoj It could work if the calculated unheldTotal is also in the iou currency

**Comment by bernhardoj** [comment]
Yes, the unheld total is also in the IOU currency (the green button).

https://github.com/user-attachments/assets/8ac23fc4-8ec4-49d5-b7b4-e0a31f79a268

The approve confirmation modal uses `unheldTotal` and `total`. Since both `total` and `unheldTotal` are in the IOU currency, I think we can safely use them. 

But I honestly still don't understand about this [one](https://github.com/Expensify/App/pull/49279#issuecomment-2353428039)

**Comment by s77rt** [comment]
@bernhardoj Regarding https://github.com/Expensify/App/pull/49279#issuecomment-2353428039, I think what I reproduced was an edge-case. Basically what I was trying to say is that we have 2 totals, `total` and `nonReimbursableTotal` and the total that was being calculated was not clear which one of those (or if it's the sum of the two). After looking into the code I can't find much about non-reimbursable transactions and I think it's not implemented in ND yet. I have asked in [Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1727645640726369).

In OD however you can create non-reimbursable transactions and after testing I got a broken behaviour already, green button showing to approve 0.00 which is not correct and the total is negative and also not correct. Assuming those are not regression from the previous PR, we can probably proceed with your suggested solution.


https://github.com/user-attachments/assets/863d43d3-c1f0-4deb-81a2-68a816674fda

**Comment by bernhardoj** [comment]
> total and nonReimbursableTotal and the total that was being calculated was not clear which one of those (or if it's the sum of the two).

If we see this code, we can infer that `nonReimbursableTotal` is part of `total`.
https://github.com/Expensify/App/blob/d18766580c7e35e439ff933f3097366e36c6ceca/src/libs/ReportUtils.ts#L2719-L2729

> green button showing to approve 0.00

Since the green button shows the amount from `unheldTotal`, this means the `unheldTotal` is 0.

> the total is negative and also not correct

Expense report total is negative.

**Comment by robertjchen** [comment]
To clarify, this incorrect total calculation is only happening for the optimistic/offline action, is that correct?

**Comment by bernhardoj** [comment]
@robertjchen I'm guessing you are talking about this [one](https://github.com/Expensify/App/issues/48760#issuecomment-2376999677). The response from PayMoneyRequest doesn't return the IOU/expense report `total`, so I can't say if the BE calculation is correct or not, but yes, the total calculation currently is incorrect optimistically.

**Comment by s77rt** [comment]
@bernhardoj For now can you please raise a PR to use the `unheldTotal` to calculate the new total to fix the regression.

**Comment by bernhardoj** [comment]
@s77rt PR is ready

**Comment by RachCHopkins** [comment]
Active discussion ongoing.

**Comment by RachCHopkins** [comment]
@robertjchen I think this is waiting on your input?

**Comment by muttmuure** [comment]
After some discussion about issues that should go in #quality, we are refocusing the channel on flakiness and reliability. This is a reproducible bug with IOU reports, so it is going into #expense.

**Comment by muttmuure** [comment]
@s77rt asked @robertjchen: 

> The total (and unheldTotal) of an IOU report should always be positive even if the owner and the manager are swapped, e.g.:
> User A requests $5.
> => Total: $5, Owner: A, Manager: B
> User B requests $7.
> => Total: $2, Owner: B, Manager: A
> The bug is that the returned total is -$2
>
> Should we move forward with the PR as is and handle this separately? cc

**Comment by s77rt** [comment]
Still on it, will try get this merged in the next couple days

**Comment by RachCHopkins** [comment]
Hey @bernhardoj have you already been paid [here](https://github.com/Expensify/App/issues/48760#issuecomment-2376011292)? I assume no, since there is no additional comment from the payer, but I just want to be 100% sure.

**Comment by bernhardoj** [comment]
Yes, I've been paid before, but then we realize there is a case we don't handle correctly. But I think it would be fair if the price is increased here since we tackle multiple other issues too on the 2nd PR.

cc: @s77rt

**Comment by s77rt** [comment]
True that multiple issues have been handled but also some regressions ([1](https://github.com/Expensify/App/issues/48760#issuecomment-2376999677), [2](https://github.com/Expensify/App/issues/53425)) has been introduced. I think both would cancel each other and issue no raise and no penalties either. Keeping this at a flat $250.

**Comment by bernhardoj** [comment]
The first one isn't a regression. Before the PR, we just took 1 transaction amount from the expense, and then we calculated it all on the 1st PR, but we didn't handle the case when there was a different currency. The 2nd one as explained the root [here](https://github.com/Expensify/App/issues/53425#issuecomment-2517291496), it was done as an attempt to fix other issues too.

**Comment by s77rt** [comment]
@bernhardoj Right, the first one was already broken but the second is something that I expected to be handled in the PR. But I agree that the work turned out more complex than originally expected. @RachCHopkins Can you please bump this to $500?

As for the checklist, it has been already provided in https://github.com/Expensify/App/issues/48760#issuecomment-2376939308 but now we need to add a regression test:

## Regression Test Proposal
### Precondition:

- Has a workspace (admin)
- Workspace has approval workflow

### Test:
1. Open the workspace chat
2. Submit 2 expenses with different currencies (e.g. $10 and €5)
3. Track 1 expense (e.g. $20)
4. Hold the 2 submitted expenses
5. Approve the report partially
6. Verify the workspace chat LHN row shows the correct owed amount that is the total of the hold expenses (e.g. $15.28)

**Comment by RachCHopkins** [comment]
@robertjchen do you agree that this is worth $500 due to the extra work involved? 

And @bernhardoj and @s77rt who has been paid what so far? As far as I'm aware, @bernhardoj has been paid $250 and @s77rt has not put any payment request in yet. Is that correct?

**Comment by bernhardoj** [comment]
Yes, I've been paid $250 before.

**Comment by robertjchen** [comment]
@RachCHopkins Agreed with the increase due to the efforts involved here 👍

**Comment by s77rt** [comment]
Thank you! Can you please update the payment summary above so the approver sees the correct values

**Comment by RachCHopkins** [comment]
@s77rt [summary](https://github.com/Expensify/App/issues/48760#issuecomment-2544997766) is up to date now!

**Comment by bernhardoj** [comment]
Thanks!

Requested in ND. (again)

**Comment by JmillsExpensify** [comment]
$500 approved for @s77rt

**Comment by JmillsExpensify** [comment]
Additional $250 approved for @bernhardoj
